primedev/mods/modweaponvars.cpp

```diff
@@ -0,0 +1,311 @@
+
+#include "client/weaponx.h"
+#include "server/weaponx.h"
+#include "modweaponvars.h"
+#include "squirrel/squirrel.h"
+#include "client/r2client.h"
+#include "server/r2server.h"
+
+typedef bool (*calculateWeaponValuesType)(int bitfield, void* weapon, void* weaponVarLocation);
+typedef char* (*get2ndParamForRecalcModFuncType)(void* weapon);
+
+template <ScriptContext context> get2ndParamForRecalcModFuncType get2ndParamForRecalcModFunc;
+template <ScriptContext context> calculateWeaponValuesType _CalculateWeaponValues;
+template <ScriptContext context> WeaponVarInfo* weaponVarArray;
+template <ScriptContext context> std::vector<SQObject> weaponModCallbacks;
+
+template <ScriptContext context> bool IsWeapon(void** ent)
+{
+	if (context == ScriptContext::SERVER)
+		return *ent == CWeaponX_vftable;
+	else
+		return *ent == C_WeaponX_vftable;
+}
+
+AUTOHOOK_INIT()
+
+// Name might be wrong, we're going off of frequency of calls
+AUTOHOOK(Cl_WeaponTick, client.dll + 0x59D1E0, bool, __fastcall, (C_WeaponX * weapon))
+{
+	SQObject* entInstance = g_pSquirrel<ScriptContext::CLIENT>->__sq_createscriptinstance(weapon);
+	g_pSquirrel<ScriptContext::CLIENT>->Call("CodeCallback_PredictWeaponMods", entInstance);
+
+	bool result = Cl_WeaponTick(weapon);
+
+	return result;
+}
+
+// Name might be wrong, we're going off of frequency of calls
+AUTOHOOK(CPlayerSimulate, server.dll + 0x5A6E50, bool, __fastcall, (CBasePlayer * player, int unk_1, bool unk_2))
+{
+	SQObject* entInstance = g_pSquirrel<ScriptContext::SERVER>->__sq_createscriptinstance(player);
+
+	// we have to do it twice, before AND after, or mispredictions happen.
+	// no idea why, unfortunately :/
+	g_pSquirrel<ScriptContext::SERVER>->Call("CodeCallback_DoWeaponModsForPlayer", entInstance);
+	bool result = CPlayerSimulate(player, unk_1, unk_2);
+	g_pSquirrel<ScriptContext::SERVER>->Call("CodeCallback_DoWeaponModsForPlayer", entInstance);
+
+	return result;
+}
+
+AUTOHOOK(Cl_CalcWeaponMods, client.dll + 0x3CA0B0, bool, __fastcall, (int mods, char* unk_1, char* weaponVars, bool unk_3, int unk_4))
+{
+	bool result;
+	if (IsWeapon<ScriptContext::CLIENT>((void**)(weaponVars - 0x1700)))
+	{
+		SQObject* entInstance = g_pSquirrel<ScriptContext::CLIENT>->__sq_createscriptinstance((void**)(weaponVars - 0x1700));
+		result = Cl_CalcWeaponMods(mods, unk_1, weaponVars, unk_3, unk_4);
+		g_pSquirrel<ScriptContext::CLIENT>->Call("CodeCallback_ApplyModWeaponVars", entInstance);
+	}
+	else
+		result = Cl_CalcWeaponMods(mods, unk_1, weaponVars, unk_3, unk_4);
+
+	return result;
+}
+
+AUTOHOOK(Sv_CalcWeaponMods, server.dll + 0x6C8B80, bool, __fastcall, (int unk_0, char* unk_1, char* unk_2, bool unk_3, int unk_4))
+{
+	bool result = Sv_CalcWeaponMods(unk_0, unk_1, unk_2, unk_3, unk_4);
+
+	if (result && IsWeapon<ScriptContext::SERVER>((void**)(unk_2 - 0x1410)))
+	{
+		SQObject* entInstance = g_pSquirrel<ScriptContext::SERVER>->__sq_createscriptinstance((void**)(unk_2 - 0x1410));
```

Comment:
these offsets into `unk_2` seam weird but they should still be a struct somehow imo.

Reply:
atleast put the offset into a named constant.
or add the field to the `CWeaponX` class and get it's offset from that.

Reply:
Its already in the weapon structs, will convert into an `offsetof` tomorrow